compiler/src/Language/Bond/Codegen/Java/Class_java.hs

```diff
@@ -37,23 +228,157 @@ package #{javaPackage};
+    }
+
+    #{bondTypeStaticVariableDecl}
+
+    public static void initializeBondType() {
+        __StructBondTypeImpl.__StructBondTypeBuilderImpl.register();
+    }
+
+    static {
+        initializeBondType();
     }
+    #{bondTypeDescriptorInstanceVariableDecl}
 
-    #{doubleLineSep 1 publicField structFields}
+    #{doubleLineSep 1 publicFieldDecl structFields}
+
+    #{publicConstructorDecl}
+
+    @Override
+    public com.microsoft.bond.StructBondType<? extends com.microsoft.bond.BondSerializable> getBondType() {
+        return #{getBondTypeReturnValue};
+    }
 }|]
             where
                 interface = [lt| implements com.microsoft.bond.BondSerializable|]
                 params = angles $ sepBy ", " paramName declParams
+                declParamsWithIndex = zip [0 :: Int ..] declParams
                 baseClass x = [lt| extends #{javaType x}|]
-                javaDefault = Java.defaultValue java
-                publicField Field {..} = [lt|public #{javaType fieldType} #{fieldName} = #{javaDefault fieldType fieldDefault};|]
+                publicFieldDecl Field {..} = [lt|public #{javaType fieldType} #{fieldName};|]
+                fieldDescriptorFieldDecl Field {..} = [lt|private #{structFieldDescriptorTypeName java fieldType} #{fieldName};|]
+                fieldDescriptorFieldInit Field {..} = [lt|this.#{fieldName} = new #{structFieldDescriptorTypeName java fieldType}(#{initMethodParams});|]
+                        where
+                            initMethodParams = [lt|this#{fieldTypeParam}, #{fieldOrdinal}, "#{fieldName}", #{modifierConstantName fieldModifier}#{fieldDefaultValueParam}|]
+                                where
+                                    fieldTypeParam = if isGenericStructFieldDescriptor fieldType
+                                        then [lt|, #{structFieldDescriptorInitTypeExpr java fieldType}|]
+                                        else mempty
+                                    fieldDefaultValueParam = if not (isPrimitiveBondType fieldType) || (fieldDefault == Nothing)
+                                        then mempty
+                                        else [lt|, #{fieldDefaultValueInitParamExpr java fieldType fieldDefault}|]
+                initializeBaseAndFieldsBaseParam = structBaseDescriptorInitStructExpr java structBase
+                initializeBaseAndFieldsFieldsParam = if null structFields
+                    then mempty
+                    else [lt|, #{sepBy ", " structFieldReference structFields}|]
+                    where
+                        structFieldReference Field {..} = [lt|this.#{fieldName}|]
+                buildNewInstanceMethodBody = if null declParams
+                    then [lt|return new __StructBondTypeImpl();|]
+                    else [lt|return new __StructBondTypeImpl(new com.microsoft.bond.GenericTypeSpecialization(genericTypeArguments));|]
+                structBondTypeImplConstructorDecl = if null declParams
```

Comment:
We haven't formalized this, but I understand `*Decl` to be a declaration or a declaration+definition - `public int i;` or `public int i = 0;`. This is just a name.